src/callbacks_step/analysis.jl

```diff
@@ -184,175 +188,208 @@ function (analysis_callback::AnalysisCallback)(integrator)
-      end
-
-      for level = max_level:-1:min_level+1
-        mpi_println(" ├── level $level:    " * @sprintf("% 14d", count(isequal(level), levels)))
-      end
-      mpi_println(" └── level $min_level:    " * @sprintf("% 14d", count(isequal(min_level), levels)))
-    end
+    print_amr_information(integrator.opts.callback, mesh, solver, cache)
     mpi_println()
 
     # Open file for appending and store time step and time information
     if mpi_isroot() && analysis_callback.save_analysis
       io = open(joinpath(analysis_callback.output_directory, analysis_callback.analysis_filename), "a")
-      mpi_isroot() && @printf(io, "% 9d", iter)
-      mpi_isroot() && @printf(io, "  %10.8e", t)
-      mpi_isroot() && @printf(io, "  %10.8e", dt)
-    end
-
-    # the time derivative can be unassigned before the first step is made
-    if t == integrator.sol.prob.tspan[1]
-      du_ode = similar(integrator.u)
+      @printf(io, "% 9d", iter)
+      @printf(io, "  %10.8e", t)
+      @printf(io, "  %10.8e", dt)
     else
-      du_ode = get_du(integrator)
+      io = devnull
     end
+
+    # Calculate current time derivative (needed for semidiscrete entropy time derivative, residual, etc.)
+    du_ode = first(get_tmp_cache(integrator))
     @notimeit timer() rhs!(du_ode, integrator.u, semi, t)
     GC.@preserve du_ode begin
       du = wrap_array(du_ode, mesh, equations, solver, cache)
+      l2_error, linf_error = analysis_callback(io, du, u, integrator.u, t, semi)
+    end # GC.@preserve du_ode
 
-      # Calculate and print derived quantities (error norms, entropy etc.)
-      # Variable names required for L2 error, Linf error, and conservation error
-      if any(q in analysis_errors for q in
-            (:l2_error, :linf_error, :conservation_error, :residual))
-        mpi_print(" Variable:    ")
-        for v in eachvariable(equations)
-          mpi_isroot() && @printf("   %-14s", varnames(cons2cons, equations)[v])
-        end
-        mpi_println()
-      end
+    mpi_println("─"^100)
+    mpi_println()
 
-      # Calculate L2/Linf errors, which are also returned by analyze_solution
-      l2_error, linf_error = calc_error_norms(u, t, analyzer, semi)
+    # Add line break and close analysis file if it was opened
+    if mpi_isroot() && analysis_callback.save_analysis
+      println(io::IOStream)
+      close(io::IOStream)
+    end
+  end
 
-      # L2 error
-      if :l2_error in analysis_errors
-        mpi_print(" L2 error:    ")
-        for v in eachvariable(equations)
-          mpi_isroot() && @printf("  % 10.8e", l2_error[v])
-          analysis_callback.save_analysis && mpi_isroot() && @printf(io, "  % 10.8e", l2_error[v])
-        end
-        mpi_println()
+  # avoid re-evaluating possible FSAL stages
+  u_modified!(integrator, false)
+
+  # Return errors for EOC analysis
+  return l2_error, linf_error
+end
+
+
+function (analysis_callback::AnalysisCallback)(io, du, u, u_ode, t, semi)
```

Comment:
A docstring/comment would be helpful to indicate when this method should be called and how

Reply:
Added in ccf9d383.

Reply:
Perfect, thanks!